bblfsh/pyuast.cc

```diff
@@ -1,272 +1,760 @@
+      0,                              // tp_as_mapping
+      0,                              // tp_hash
+      0,                              // tp_call
+      0,                              // tp_str
+      0,                              // tp_getattro
+      0,                              // tp_setattro
+      0,                              // tp_as_buffer
+      Py_TPFLAGS_DEFAULT,             // tp_flags
+      "External UAST node",           // tp_doc
+      0,                              // tp_traverse
+      0,                              // tp_clear
+      0,                              // tp_richcompare
+      0,                              // tp_weaklistoffset
+      0,                              // tp_iter: __iter()__ method
+      0,                              // tp_iternext: next() method
+      PyNodeExt_methods,              // tp_methods
+      0,                              // tp_members
+      0,                              // tp_getset
+      0,                              // tp_base
+      0,                              // tp_dict
+      0,                              // tp_descr_get
+      0,                              // tp_descr_set
+      0,                              // tp_dictoffset
+      0,                              // tp_init
+      PyType_GenericAlloc,            // tp_alloc
+      0,                              // tp_new
+    };
 }
 
-static const char *Token(const void *node) {
-  return String(node, "token");
-}
+// ==========================================
+//          External UAST iterator
+// ==========================================
 
-static size_t ChildrenSize(const void *node) {
-  return Size(node, "children");
-}
+typedef struct {
+  PyObject_HEAD
+  ContextExt *ctx;
+  uast::Iterator<NodeHandle> *iter;
+  bool freeCtx;
+} PyUastIterExt;
 
-static void *ChildAt(const void *node, int index) {
-  PyObject *children = AttributeValue(node, "children");
-  void *retval = nullptr;
-  if (children) {
-    retval = ItemAt(children, index);
-    Py_DECREF(children);
-  }
+static void PyUastIterExt_dealloc(PyObject *self);
 
-  return retval;
+static PyObject *PyUastIterExt_iter(PyObject *self) {
+  Py_INCREF(self);
+  return self;
 }
 
-static size_t RolesSize(const void *node) {
-  return Size(node, "roles");
-}
+static PyObject *PyUastIterExt_toPy(ContextExt *ctx, NodeHandle node);
 
-static uint16_t RoleAt(const void *node, int index) {
-  uint16_t retval = 0;
-  PyObject *roles = AttributeValue(node, "roles");
-  if (roles) {
-    retval = (uint16_t)PyLong_AsUnsignedLong(ItemAt(roles, index));
-    Py_DECREF(roles);
-  }
-  return retval;
-}
+static PyObject *PyUastIterExt_next(PyObject *self) {
+  auto it = (PyUastIterExt *)self;
 
-static size_t PropertiesSize(const void *node) {
-  size_t retval = 0;
-  PyObject *properties = AttributeValue(node, "properties");
-  if (properties) {
-    retval = PyMapping_Size(properties);
-    Py_DECREF(properties);
+  // TODO: check errors
```

Comment:
Is this still relevant and if it is, could it affect the correctness of the result?